### DISCUSSION TITLE
Leaf-node -> cluster connection pooling plans and simulation

### DISCUSSION THREAD

**Discussion by AlexeiSosnov** [nats-io/nats-server#4704] [Q&A] [answered]
Hello

To my understanding, there is no connection pooling between leaf and upstream hub nodes; only single connection is used. Having in mind there is no [visible] way to set delivery priority for mirror / source streams or specific subjects in streams, there is a risk that high-volume "low-business-priority" traffic will congest low-volume "high-business-priority" message or request-reply command traffic. It can lead to command timeouts and slow delivery of important messages between hub and leaf nodes.

So the questions:
- are there any plans to introduce connection pooling (with account traffic separation, etc) between leaf and hub nodes, like it is already implemented for cluster nodes? If yes, are there any time estimations (2024, 2025, etc.)?
- if no, are PRs could be accepted for such a feature?
- is it possible to "simulate" leaf-nodes with connection pooling by a) creating usual single-node NATS clusters on leaf nodes, b) grouping those NATS cluster with the hub NATS cluster in a super-cluster, using v2 routes, c) placing streams on leaf and hub nodes by tags, d) using different accounts on "leaf" and "hub" clusters for traffic access control?

**Comment by derekcollison** [comment, marked as answer]
Each leafnode server will have a single connection per account. You can have multiple connections based on different local accounts, so you could separate traffic that way as needed.
